src/modules/block/index.ts

```diff
@@ -0,0 +1,308 @@
+    FLOW_CDU: 'flow:component-did-update',
+    FLOW_RENDER: 'flow:render',
+  };
+
+  children: Children;
+
+  eventBus: () => EventBus<Events>;
+
+  protected readonly props: P;
+
+  private _template: string;
+
+  private _id: Nullable<string>;
+
+  private _element: Nullable<HTMLElement> = null;
+
+  constructor(tmpl: string, propsAndChildren: Record<string, Block | any>) {
+    const eventBus = new EventBus();
+    const { children, props } = this._getChildren(propsAndChildren);
+
+    this.children = children;
+
+    this._template = tmpl;
+
+    this._id = makeUUID();
+
+    this.props = this._makePropsProxy(props);
+
+    this.eventBus = () => eventBus;
+
+    this._registerEvents(eventBus);
+    eventBus.emit(Block.EVENTS.INIT);
+  }
+
+  _registerEvents(eventBus: EventBus<Events>) {
+    eventBus.on(Block.EVENTS.INIT, this.init.bind(this));
+    eventBus.on(Block.EVENTS.FLOW_CDM, this._componentDidMount.bind(this));
+    eventBus.on(Block.EVENTS.FLOW_CDU, this._componentDidUpdate.bind(this));
+    eventBus.on(Block.EVENTS.FLOW_RENDER, this._render.bind(this));
+  }
+
+  init() {
+    this.eventBus().emit(Block.EVENTS.FLOW_RENDER);
+  }
+
+  _componentDidMount() {
+    this.componentDidMount();
+
+    Object.values(this.children).forEach((child) => {
+      if (Array.isArray(child)) {
+        child.forEach((subChild) => {
+          if (this._isBlocksObject(subChild)) {
+            Object.values(subChild).forEach((el) => {
+              el.dispatchComponentDidMount();
+            });
+          } else {
+            (subChild as Block).dispatchComponentDidMount();
+          }
+        });
+      } else {
+        child.dispatchComponentDidMount();
+      }
+    });
+  }
+
+  componentDidMount(oldProps?: P) {
+    this.eventBus().emit(Block.EVENTS.FLOW_CDU, oldProps, this.props);
+  }
+
+  dispatchComponentDidMount() {
+    this.eventBus().emit(Block.EVENTS.FLOW_CDM);
+  }
+
+  _componentDidUpdate(oldProps: P, newProps: P) {
+    const response = this.componentDidUpdate(oldProps, newProps);
+
+    if (response) {
+      this.eventBus().emit(Block.EVENTS.FLOW_RENDER);
+    }
+  }
+
+  componentDidUpdate(oldProps: P, newProps: P) {
+    if (!oldProps) {
+      return true;
+    }
+
+    const updatedProp = Object.keys(newProps).find(
+      (prop) =>
+        (oldProps as Record<string, any>)[prop] === undefined ||
+        (oldProps as Record<string, any>)[prop] !==
+          (newProps as Record<string, any>)[prop],
+    );
+
+    return !!updatedProp;
+  }
+
+  setProps = (nextProps: Partial<P>) => {
+    if (!nextProps) {
+      return;
+    }
+
+    Object.assign(this.props, nextProps);
+  };
+
+  get element() {
+    return this._element;
+  }
+
+  _addEvents() {
+    const { events = {} } = this.props as unknown as Record<string, any>;
+
+    if (!this._element) {
+      return;
+    }
+
+    Object.keys(events).forEach((eventName) => {
+      this._element!.addEventListener(eventName, events[eventName]);
+    });

```

Comment:
Здесь тоже. А вообще здесь можно сократить
```suggestion
    Object.keys(events).forEach((eventName) => {
      this._element?.addEventListener(eventName, events[eventName]);
    });
```